server/src/main/java/org/opensearch/common/blobstore/transfer/stream/ResettableCheckedInputStream.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.blobstore.transfer.stream;
+
+import com.jcraft.jzlib.CRC32;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+
+/**
+ * ResettableCheckedInputStream is a modified implementation of {@link java.util.zip.CheckedInputStream} that supports
+ * mark and reset and modifies the file checksum during mark and reset calls.
+ *
+ * @opensearch.internal
+ */
+public class ResettableCheckedInputStream extends FilterInputStream {
+    private final CRC32 checksum;
+    private final CRC32 markedChecksum;
+    private final long startPos;
+    private final String file;
+    private final OffsetRangeInputStream offsetRangeInputStream;
+
+    /**
+     * Creates a new ResettableCheckedInputStream object
+     *
+     * @param offsetRangeInputStream   The underlying input stream
+     * @param file Name of the file
+     */
+    public ResettableCheckedInputStream(OffsetRangeInputStream offsetRangeInputStream, String file) throws IOException {
```

Comment:
I believe that this should be explicitly set to true/false in every stream otherwise it can lead to inconsistent behavior. Lets say there is leaf stream which doesn't supports mark and also doesn't override markSupported but parent stream supports mark and overrides markSupported to true. In this case it will lead to a faulty scenario as now reset/mark are available for use even though stream doesn't support it.

Reply:
One such bug is present in an internal class CheckedInputStream . Stream doesn't support mark and reset because it works with irreversible checksums and also doesn't override markSupported. If this stream is wrapped over FileInputStream then any retry during the entire stream read will lead to invalid checksum and consequently, false CorruptedException errors.